### ISSUE TITLE
ACK Via header - incorrect branch tag

### ISSUE THREAD

**Issue by l7s** [onsip/SIP.js#517]
Hi,

There is a bug in v. 0.9.2 which results in `branch` tag in ACK request sent in response to `SIP/2.0 407 Proxy Authentication Required` being different string to `branch` tag in initial INVITE. This is against RFC 3261, as  `branch` tag in Via header should be identical to initial INVITE. To reproduce the problem:

1. Place a call to a SIP proxy which challenges INVITES with `407` response.
2. ACK response will be generated by SIPJS.

expected behaviour:
`branch` tag in Via header of ACK response is identical to initial INVITE

actual behaviour:
`branch` tag in Via header of ACK response is different to initial INVITE

example call flow:
```
INVITE sip:123@example.net SIP/2.0
Via: SIP/2.0/WSS 31a44ueqgapu.invalid;branch=z9hG4bK9560030

SIP/2.0 407 Proxy Authentication Required

ACK sip:123@example.net SIP/2.0
Via: SIP/2.0/WSS 31a44ueqgapu.invalid;branch=z9hG4bK3227832
```

Note: the above bug didn't exist in v. 0.7.7.

Regards,
Chris

**Comment by l7s** [comment]
Hi, below further information on why current behaviour is broken:

https://andrewjprokop.wordpress.com/2014/03/06/understanding-the-sip-via-header/

"The branch parameter must be unique across space and time for all requests sent by a user agent.  The exceptions to this rule are CANCEL and **ACK for non-2xx responses**.  A CANCEL will have the same branch parameter as the request it cancels.  **An ACK for a non-2xx response will have the same branch parameter as the INVITE whose response it acknowledges**."

... I am trying to fix this, but any hints on how to best approach the issue much appreciated.

Regards,
Chris

**Comment by egreenmachine** [comment]
Hi Chris,

So there are a few bugs with the way that SIP.js is generating ACK's at the moment. The underlying issue is that the Transaction, Dialog, and Request layers need a bit of a refactor to properly track and handle this stuff. Currently, most of the responsibility ends up in the session layer, which is the wrong place for this (same with 100 rel stuff, but that is somewhat unrelated to what you are asking). The reason that it is working in 0.7.X and not 0.8.X is that I tried to move the ACK handling into the Transaction layer, which is where all of these issues started. 55cb0aeff36d314e20292dbf8fb1ddbc0f3cd7b0 is the commit in question. The reason I did this, was to help with retransmissions. The Transaction layer, should silently handle the retransmissions according to the SIP RFC, and the Session and Dialog layers should be unaware of it. Unfortunately, it is a big mess right now. So... the best approach is to go and fix all of this. The plan from my end is to start with the Transaction layer. Make it better at actually tracking the Transactions. Then fix the Dialog layer, then the Session layer. It is a big project, and unfortunately there are more pressing matters which is why it remains unresolved. 

The simplest option for you, may be to try and back out the commit referenced above. Use the request sender to generate new ACK's where you need it. It will not be all that straightforward, but it would be simpler then trying to fix the underlying problem.

Hope all of that helps. Let me know if I can be of further assistance.

**Comment by l7s** [comment]
Hi Eric,

Many thanks for your reply. For now I did a "quick hack" to make this work correctly:

```
diff --git a/sip-0.9.2.js b/sip-0.9.2.js
index cbb98cf..4d8e980 100644
--- a/sip-0.9.2.js
+++ b/sip-0.9.2.js
 /***/ }),
@@ -1860,6 +1860,7 @@ module.exports = function (SIP) {
     this.extraHeaders = (extraHeaders || []).slice();
     this.statusCode = params.status_code;
     this.reasonPhrase = params.reason_phrase;
+    this.branchTag = params.branchTag;
 
     // Fill the Common SIP Request Headers
 
@@ -2958,6 +2959,8 @@ module.exports = function (SIP) {
     // TODO: Move PRACK stuff into the transaction layer. That is really where it should be
 
     var self = this,
+        dialogId = this.response.call_id + this.response.from_tag + this.response.to_tag,
+        route_set = (this.request.ua.dialogs[dialogId]) ? this.request.ua.dialogs[dialogId].route_set : this.response.getHeaders('record-route').reverse(),
         ruri;
     options = options || {};
 
@@ -2973,7 +2976,7 @@ module.exports = function (SIP) {
       from_tag: this.response.from_tag,
       to_uri: this.response.to.uri,
       to_tag: this.response.to_tag,
-      route_set: this.response.getHeaders('record-route').reverse()
+      route_set: route_set
     }, options.extraHeaders || [], options.body);
 
     this.ackSender = new SIP.RequestSender({
@@ -3069,7 +3072,7 @@ module.exports = function (SIP) {
         case C.STATUS_CALLING:
         case C.STATUS_PROCEEDING:
           this.stateChanged(C.STATUS_COMPLETED);
-          this.sendACK();
+          this.sendACK({ branchTag: tr.id });
           this.request_sender.receiveResponse(response);
           break;
         case C.STATUS_COMPLETED:
@@ -3090,7 +3093,7 @@ module.exports = function (SIP) {
     var via;
 
     this.transport = transport;
-    this.id = 'z9hG4bK' + Math.floor(Math.random() * 10000000);
+    this.id = (request.branchTag) ? request.branchTag : 'z9hG4bK' + Math.floor(Math.random() * 10000000);
     this.request_sender = request_sender;
     this.request = request;
 ````

above also fixes issue reported in #506 

I am more than happy to allocate resources to help moving forward re-factoring you described. If you wanted to discuss please email me at chris at ssl7.net

Best regards,
Chris
